### DISCUSSION TITLE
Both library and client application _internally_ using spdlog, OK?

### DISCUSSION THREAD

**Discussion by N-Dekker** [gabime/spdlog#2622] [Q&A]
Is it possible for a static library to internally use spdlog (header-only), and then have a client application linking to the static library and using another instance of spdlog as well?

I see this is similar to discussion https://github.com/gabime/spdlog/discussions/1937 started by @simone-gaia but there is a main difference: "my" static library does not intend to share its internal spdlog instance with the client application. (So clearly, my static library has no spdlog header #include's in its own header files.) The client application should just use its own spdlog instance. Would that be OK?

**Comment by N-Dekker** [comment]
When using the spdlog as header-only internally inside my static library, the following seems to do the job, inside my C++ (.cpp/.cxx) source file:

    #define spdlog  spdlog_instance_of_my_library

    // Do only #include header files from gabime/spdlog _after_ the #define
    #include <spdlog/spdlog.h>
    #include <spdlog/sinks/basic_file_sink.h>
    ...


Would that be just fine?

**Comment by tt4g** [comment]
The same answer as #1937.
If the macros and compiler settings related to spdlog are the same in the two libraries, there is no problem since they are binary compatible.
Which library symbols are used is determined by the order in which they are passed to the linker. The first symbols found by the linker are included in the application binary, while those found later are ignored.

**Reply by N-Dekker** [reply]
Thanks @tt4g but renaming the `spdlog` namespace by a `#define spdlog spdlog_instance_of_my_library` (as in https://github.com/gabime/spdlog/discussions/2622#discussioncomment-4780603) does effectively change the name of all spdlog library symbols as seen by the linker, right?

**Reply by tt4g** [reply]
Probably yes.
However, this behavior depends on the behavior of how the compiler mangles the name.
If the symbols are different, theoretically no symbol collision can occur.

Note that the application size will be larger because the spdlog binaries of the two static libraries are included in the application.
And since spdlog bundles the fmt library, the symbols in the fmt library are considered to be in conflict.

**Reply by N-Dekker** [reply]
Thanks, yes indeed, the symbols of the fmt library might still cause a conflict that way. Fortunately there appears a solution here: https://github.com/fmtlib/fmt/issues/2642  It appears that a user may redefine `FMT_BEGIN_NAMESPACE`! https://github.com/gabime/spdlog/blob/5a589438d2d2abc67bdbec2aad7e2a94800c00ea/include/spdlog/fmt/bundled/core.h#L213-L220 Hope that will work while using spdlog as well 🤔
